Add a simple frame buffer control to the ddr_access_wrapper.

minhtue1996 opened the pull request:
Both DDR writer and reader HLS modules take in a base address in DDR for write and read.  Previously the base addresses are always tied to the same value, so writer and reader are accessing the same frame simultaneously.  This can cause the read side to read a "frame" that contains pixels from different frames (because a new frame is being written while read is happening).

This commit adds a simple FrameBufferControl module such that the writer and reader will access different frames.
The FrameBufferControl module provides the base addresses for the writer and reader, and updates the addresses when the writer finishes writing a frame (takes writer_finish as input).
Each time the DDR writer finishes writing a frame to DDR, FrameBufferControl provides a new address for the DDR writer to write the next frame, and provides to the DDR reader with the address of latest frame (the frame that just got written to DDR by DDR writer).
The number of frames and each frame's byte size can be parameterized.

Also do some refactoring to `debayer.hpp` and `format_conversions.hpp`.

minhtue1996 commented:
Testing:
Run cosim locally using the 3 interfaces i.e.
- RAM_INTERFACE=1 shls -a cosim
- AXI4_S_INTERFACE=1 shls -a cosim
- AXI4_STREAM_INTERFACE=1 shls -a cosim
All 3 works.

Some points of discussion:
1. Should the macro name be changed? Right now I make it match the slide since I think this is the terminology we'll tell the customers, but it can be confusing between AXI4_S and AXI4_STREAM. Maybe something like AXI4_TARGET and AXI4_STREAM is better.
2. When I add the dg.exp to the main repo, aside from running cosim, what else should I test for? I'm thinking of the following:
- Cycle latency
- II
- Interface report

LannyLianMCHP commented:
> Testing: Run cosim locally using the 3 interfaces i.e.
> 
> * RAM_INTERFACE=1 shls -a cosim
> * AXI4_S_INTERFACE=1 shls -a cosim
> * AXI4_STREAM_INTERFACE=1 shls -a cosim
>   All 3 works.
> 
> Some points of discussion:
> 
> 1. Should the macro name be changed? Right now I make it match the slide since I think this is the terminology we'll tell the customers, but it can be confusing between AXI4_S and AXI4_STREAM. Maybe something like AXI4_TARGET and AXI4_STREAM is better.
> 2. When I add the dg.exp to the main repo, aside from running cosim, what else should I test for? I'm thinking of the following:
> 
> * Cycle latency
> * II
> * Interface report

1. Let's have the MACROs match the pragma names, i.e., AXI_TARGET, AXI_INITIATOR, and for FIFO: use AXI_STREAM
2. The main check is interface report, making sure the desired ports exist for the specific interfaces.

minhtue1996 commented:
I updated the PR with the testbench rewritten. Note that this only works with struct-field pragma support (see PR-644 of the main repo).

minhtue1996 committed:
Add a simple frame buffer control to the ddr_access_wrapper.

Both DDR writer and reader HLS modules take in a base address in DDR for write
and read.  Previously the base addresses are always tied to the same value,
so writer and reader are accessing the same frame simultaneously.  This can
cause the read side to read a "frame" that contains pixels from different
frames (because a new frame is being written while read is happening).

This commit adds a simple FrameBufferControl module such that the writer and
reader will access different frames.
The FrameBufferControl module provides the base addresses for the writer and
reader, and updates the addresses when the writer finishes writing a frame
(takes writer_finish as input).
Each time the DDR writer finishes writing a frame to DDR, FrameBufferControl
provides a new address for the DDR writer to write the next frame, and
provides to the DDR reader with the address of latest frame (the frame that
just got written to DDR by DDR writer).
The number of frames and each frame's byte size can be parameterized.

Also do some refactoring to `debayer.hpp` and `format_conversions.hpp`.
